utils/yamlConfig.js

```diff
@@ -0,0 +1,146 @@
+const mergeUtils = require('./mergeUtils');
+
+function encodeKey(key) {
+  return key.replace(/[^a-zA-Z0-9]/g, char => '_x' + char.charCodeAt(0).toString(16));
+}
+
+function flattenEnvObject(obj, prefix) {
+  const result = {};
+  const path = prefix;
+  flattenEnvObjectInternal(obj, path, result);
+  return result;
+
+  function flattenEnvObjectInternal(obj, path, result) {
+    for (const key of Object.keys(obj)) {
+      const encodedKey = encodeKey(key);
+      const newPath = path ? `${path}_${encodedKey}` : encodedKey;
+      const val = obj[key];
+      if (typeof val === 'object') {
+        flattenEnvObjectInternal(val, newPath, result);
+      } else if (Array.isArray(val)) {
+        result[newPath] = val.join(',') + ',';
+      } else {
+        result[newPath] = String(val);
+      }
+    }
+  }
+}
+
+function convertConfigToEnvObj(config) {
+  return flattenEnvObject(config, 'ZWED');
+}
+
+function convertConfigToEnvSource(config) {
+  const envObj = convertConfigToEnvObj(config);
+  return Object.keys(envObj).map(key => `export ${key}="${envObj[key]}"`).join('\n');
+}
+
+function getHaInstanceId() {
+  return process.env['ZWELS_HA_INSTANCE_ID'];
+}
+
+function getInstanceDir() {
+  return process.env['INSTANCE_DIR'] || '~/.zowe';
+}
+
+function omitCommonConfigKeys(config) {
+  const commonConfigKeys = [
+    'certificate',
+    'crossMemoryServerName',
+    'enabled',
+    'launcher',
+    'port',
+    'tls'
+  ];
+  return _.omit(config, commonConfigKeys);
+}
+
+function getComponentConfig(yawn, component, haInstanceId) {
+  const componentLevelConfig = _.get(yawn.json, ['components', component]);
+  const instanceLevelConfig = _.get(yawn.json, ['haInstances', haInstanceId, 'components', component]);
+  const config = mergeUtils.deepAssign(componentLevelConfig, instanceLevelConfig ? instanceLevelConfig : {});
+  return omitCommonConfigKeys(config);
+}
+
+function getYamlConfig(yawn, haInstanceId) {
+  const appServerConfig = getComponentConfig(yawn, 'app-server', haInstanceId);
+  const zssConfig = getComponentConfig(yawn, 'zss', haInstanceId);
+  if (zssConfig && appServerConfig) {
+    return mergeUtils.deepAssign(zssConfig, appServerConfig);
+  } else if (appServerConfig) {
+    return appServerConfig;
+  } else if (zssConfig) {
+    return zssConfig;
+  }
+  return undefined;
+}
+
+function loadZoweDotYaml() {
+  const instanceDir = getInstanceDir();
+  const zoweDotYamlFile = path.join(instanceDir, 'zowe.yaml');
+  const instanceDotEnvFile = path.join(instanceDir, 'instance.env');
+  if (fs.existsSync(instanceDotEnvFile)) {
+    // instance.env is higher priority than zowe.yaml
+    return;
+  }
+  if (!fs.existsSync(zoweDotYamlFile)) {
+    // zowe.yaml not found
+    return;
+  }
+  const yawn = parseZoweDotYaml(zoweDotYamlFile);
+  return yawn;
+}
+
+function parseZoweDotYaml(zoweDotYamlFile) {
+  let yawn;
+  try {
+    const yamlText = fs.readFileSync(zoweDotYamlFile).toString();
```

Comment:
This assumes too much about the location of the configuration.
Please change this to `loadYamlConfig(yamlFile:path, instanceFile?:path)`

If the names and paths change later, it's the responsibility of the implementation code in zlux-app-server to find the right paths.

There could be forks / alternatives to zlux-app-server that want to use different paths.